### PR TITLE
OBPIH-2655: Allow editing pending purchase orders based SM

### PR DESCRIPTION
**Pull request by awalkowiak**

### CODE REVIEW

File: grails-app/controllers/org/pih/warehouse/order/OrderController.groovy [hunk 1]
```diff
@@ -83,9 +84,16 @@ class OrderController {
         }
 
         try {
+            def shipmentIds = command.order?.orderItems?.shipmentItems*.shipment*.id.flatten().unique()
```
Comment (jmiranda) [inline grails-app/controllers/org/pih/warehouse/order/OrderController.groovy]: First we need to make sure we only update PENDING shipments. 

In order to make this more DRY, let's add methods to Order and Shipment to getShipments() and getOrders(), respectively. There should be one method that gets all and another that allows you to get by status. 

Something like this. And make sure to add "shipments" to the transients list.
```
List getShipments() { 
    orderItems?.shipmentItems*.shipment*.flatten()?.unique() 
}
List getShipments(ShipmentStatusCode statusCode) { 
    orderItems?.shipmentItems*.shipment.flatten()?.unique().findAll { it.currentStatus == statusCode }
}
```
then this line would be something like this
```
def shipmentIds = command.order?.getShipments(ShipmentStatusCode.PENDING).collect { it.id }
```

Comment (awalkowiak) [reply grails-app/controllers/org/pih/warehouse/order/OrderController.groovy]: @jmiranda  Ugh,  i have not noticed that there is already listShipments() in Order. Ill fix that today.

Comment (jmiranda) [reply grails-app/controllers/org/pih/warehouse/order/OrderController.groovy]: @awalkowiak Let me know if this is ready for review again.

Comment (awalkowiak) [reply grails-app/controllers/org/pih/warehouse/order/OrderController.groovy]: @jmiranda Ready

File: grails-app/domain/org/pih/warehouse/order/Order.groovy [hunk 2]
```diff
@@ -176,10 +178,13 @@ class Order implements Serializable {
         return (status == OrderStatus.CANCELED)
     }
 
-    def listShipments() {
+    def getShipments() {
         return orderItems.collect { it.listShipments() }.flatten().unique() { it?.id }
     }
 
+    List getShipmentsByStatus(ShipmentStatusCode statusCode) {
+        return getShipments().findAll { it -> it.currentStatus == statusCode }
```
Comment (jmiranda) [inline grails-app/domain/org/pih/warehouse/order/Order.groovy]: You can just use `shipments.findAll { ... }` here.

File: grails-app/services/org/pih/warehouse/order/OrderService.groovy [hunk 3]
```diff
@@ -144,7 +144,7 @@ class OrderService {
      */
     OrderCommand saveOrderShipment(OrderCommand orderCommand) {
         def shipmentInstance = new Shipment()
-        def shipments = orderCommand?.order?.listShipments()
+        def shipments = orderCommand?.order?.getShipments()
```
Comment (jmiranda) [inline grails-app/services/org/pih/warehouse/order/OrderService.groovy]: Same here

File: grails-app/services/org/pih/warehouse/order/OrderService.groovy [hunk 4]
```diff
@@ -335,7 +335,7 @@ class OrderService {
         try {
 
             if (orderInstance.status in [OrderStatus.PLACED,  OrderStatus.PARTIALLY_RECEIVED, OrderStatus.RECEIVED]) {
-                orderInstance?.listShipments().each { Shipment shipmentInstance ->
+                orderInstance?.getShipments().each { Shipment shipmentInstance ->
```
Comment (jmiranda) [inline grails-app/services/org/pih/warehouse/order/OrderService.groovy]: And here

File: grails-app/services/org/pih/warehouse/shipping/ShipmentService.groovy [hunk 5]
```diff
@@ -2130,4 +2131,45 @@ class ShipmentService {
...
+            itemsToRemove.each { it -> deleteShipmentItem(it) }
+
+            order.orderItems.each { OrderItem orderItem ->
+                def shipmentItem = shipment.shipmentItems.find { it.orderItems?.any { it.id == orderItem.id } }
+                def lotNumber = shipmentItem?.lotNumber
+                def expirationDate = shipmentItem?.expirationDate
+                if (!shipmentItem) {
+                    shipmentItem = new ShipmentItem(
+                        product: orderItem.product,
+                        inventoryItem: orderItem.inventoryItem,
```
Comment (jmiranda) [inline grails-app/services/org/pih/warehouse/shipping/ShipmentService.groovy]: @awalkowiak `orderItem.inventoryItem` will probably be null 99% of the time, if not 100%. If there's a product and lot number specified then we need to do a look up of the InventoryItem by product and lot number. But I can't tell from the code what's going on here and whether we need to use InventoryService.findOrCreateInventoryItem() or InventoryService.findInventoryItemByProductAndLotNumber() or something else. 

But all this logic looks complicated so let's have a discuss before we move this one forward. Can you enumerate the different use cases that are represented here:
- User goes back to edit order item quantity
- User goes back to edit order item product (?)
- ...

Comment (awalkowiak) [reply grails-app/services/org/pih/warehouse/shipping/ShipmentService.groovy]: @jmiranda you're right inventory item is not required here. I removed it. Use cases:
- edit line item (qty)
- add new line item
-delete line item

File: grails-app/domain/org/pih/warehouse/order/Order.groovy (hunk 2 above)
Comment (awalkowiak) [reply grails-app/domain/org/pih/warehouse/order/Order.groovy]: Fixed

File: grails-app/services/org/pih/warehouse/order/OrderService.groovy (hunk 3 above)
Comment (awalkowiak) [reply grails-app/services/org/pih/warehouse/order/OrderService.groovy]: Fixed

File: grails-app/services/org/pih/warehouse/order/OrderService.groovy (hunk 4 above)
Comment (awalkowiak) [reply grails-app/services/org/pih/warehouse/order/OrderService.groovy]: Fixed

File: grails-app/services/org/pih/warehouse/shipping/ShipmentService.groovy (hunk 5 above)
Comment (jmiranda) [reply grails-app/services/org/pih/warehouse/shipping/ShipmentService.groovy]: @awalkowiak But if there's a lot number and expiration date, it is required. Let's discuss on Slack.

**Pull request merged**